pkg/snapshotmgr/snapshot_manager.go

```diff
@@ -691,17 +691,17 @@ func (this *SnapshotManager) CreateVolumeFromSnapshot(sourcePEID astrolabe.Prote
 		}
 		download, err = pluginClient.DatamoverV1alpha1().Downloads(veleroNs).Get(context.TODO(), downloadRecordName, metav1.GetOptions{})
 		if err != nil {
-			this.Errorf("Retrieve download record %s failed with err %v", downloadRecordName, err)
-			return false, errors.Wrapf(err, "Failed to retrieve download record %s", downloadRecordName)
+			this.Errorf("Retrieve download record %s failed with err %v, retrying..", downloadRecordName, err)
+			return false, nil
```

Comment:
How did the download succeed with the existing code in the logs?  Did it create new Download CR instead of using the original one?

Reply:
Yes, a new download record was created, and copy from remote to local happened twice.

```
Remote PE: ivd:56dffde6-bcd9-42c0-aa13-2a7b3549837b:529d6c6c-c29b-4355-b93e-3da59c5a92b5
```

```
FAILED: kibishii-plqqwjc/etcd0-pv-claim -> download-529d6c6c-c29b-4355-b93e-3da59c5a92b5-29e857f8-cbf3-4900-94a5-3c8fc130f9cc

data-manager.log

time="2021-12-15T18:36:00Z" level=info msg="Successfully Overwrote current IVD PE ivd:203c9a2c-61d4-49c7-a2fc-5883d9b403d0 with the source PE ivd:56dffde6-bcd9-42c0-aa13-2a7b3549837b:529d6c6c-c29b-4355-b93e-3da59c5a92b5" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/ivd/ivd_protected_entity.go:510"
.
.
time="2021-12-15T18:35:13Z" level=info msg="Download status updated from InProgress to Completed" controller=download generation=2 logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/controller/download_controller.go:417" name=download-529d6c6c-c29b-4355-b93e-3da59c5a92b5-29e857f8-cbf3-4900-94a5-3c8fc130f9cc namespace=velero phase=InProgress
```

As you can see data-mgr processed the download successfully, remote data was copied into `ivd:203c9a2c-61d4-49c7-a2fc-5883d9b403d0`. However, since there was a retrieve failure, it reported an error to `cloneFromSnapshot`. The `cloneFromSnapshot` was processed again.

```
backup-driver.log:

time="2021-12-15T18:35:22Z" level=info msg="PVC kibishii-plqqwjc/etcd0-pv-claim already exists, reusing" logSource="/go/pkg/mod/github.com/vmware-tanzu/astrolabe@v0.4.1-0.20210813185044-12eb18c3f6d5/pkg/pvc/pvc_protected_entity_type_manager.go:175"
```

As seen above, `backup-driver` observes that there is already an existing PVC and it reuses it. This is followed by creating a new download CR for the same remote pe.

```
time="2021-12-15T18:35:22Z" level=info msg="Start creating Download CR for ivd:56dffde6-bcd9-42c0-aa13-2a7b3549837b:529d6c6c-c29b-4355-b93e-3da59c5a92b5" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr/snapshot_manager.go:627"
time="2021-12-15T18:35:22Z" level=info msg="CloneFromSnapshotReference: kibishii-plqqwjc/4622e49c-2919-4ac3-a10d-d7512e0b7ef8" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr/snapshot_manager.go:655"
time="2021-12-15T18:35:22Z" level=info msg="Creating Download CR: velero/download-529d6c6c-c29b-4355-b93e-3da59c5a92b5-43eb7c90-29be-4dd2-b1a1-772354e6f248" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr/snapshot_manager.go:657"
time="2021-12-15T18:35:22Z" level=info msg="Ready to create download CR. Will retry on network issue every 5 seconds for 5 retries at maximum" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr/snapshot_manager.go:669"
time="2021-12-15T18:36:22Z" level=info msg="Polling download record download-529d6c6c-c29b-4355-b93e-3da59c5a92b5-43eb7c90-29be-4dd2-b1a1-772354e6f248" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr/snapshot_manager.go:689"
```

The data-manager processes this download CR and overwrites the local repo from the remote repo again:

```
PASSED: kibishii-plqqwjc/etcd0-pv-claim -> download-529d6c6c-c29b-4355-b93e-3da59c5a92b5-43eb7c90-29be-4dd2-b1a1-772354e6f248

data-manager.log:

time="2021-12-15T18:35:23Z" level=info msg="Overwriting current IVD PE ivd:203c9a2c-61d4-49c7-a2fc-5883d9b403d0 with the source PE ivd:56dffde6-bcd9-42c0-aa13-2a7b3549837b:529d6c6c-c29b-4355-b93e-3da59c5a92b5" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/ivd/ivd_protected_entity.go:482"
.
.
time="2021-12-15T18:41:04Z" level=info msg="Successfully Overwrote current IVD PE ivd:203c9a2c-61d4-49c7-a2fc-5883d9b403d0 with the source PE ivd:56dffde6-bcd9-42c0-aa13-2a7b3549837b:529d6c6c-c29b-4355-b93e-3da59c5a92b5" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/ivd/ivd_protected_entity.go:510"
.
.
time="2021-12-15T18:41:04Z" level=info msg="Download status updated from InProgress to Completed" controller=download generation=6 logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/controller/download_controller.go:417" name=download-529d6c6c-c29b-4355-b93e-3da59c5a92b5-43eb7c90-29be-4dd2-b1a1-772354e6f248 namespace=velero phase=InProgress
```

This leads to `cloneFromSnapshot` moving to a `Completed` phase implying a successful restore.

The current change will prevent this specific workflow, instead will keep retrying until it gets the download CR status is retrieved.

Reply:
> The current change will prevent this specific workflow, instead will keep retrying until it gets the download CR status is retrieved.

Does this retry forever?

Reply:
as mentioned offline, it tries until we get a valid download status (Completed/Failed), retries infinitely if we are unable to retrieve download cr.